### PR TITLE
Laura/reduce consensus sumbission latency

### PR DESCRIPTION
**Pull request by laura-makdah**

### CODE REVIEW

**Comment by asonnino** [comment]
Is this PR ready for review?

File: crates/sui-types/src/sui_system_state.rs
Comment (akichidis) [inline crates/sui-types/src/sui_system_state.rs:296]: Probably we should have a message closer to SUI? ex `Can't get network key`

File: crates/sui-types/src/sui_system_state.rs
Comment (akichidis) [inline crates/sui-types/src/sui_system_state.rs:315]: I am not sure we'll need those - a validator won't attempt on its own to connect to any worker node

File: crates/sui-types/src/sui_system_state.rs
Comment (akichidis) [inline crates/sui-types/src/sui_system_state.rs:302]: `other_primary ---> other_validator` ?

File: crates/sui-types/src/sui_system_state.rs
Comment (akichidis) [inline crates/sui-types/src/sui_system_state.rs:286]: I am not 100% sure this method gives additional value in the SUI validator case, as a validator will just try to connect to other validators (we don't have the same distinction between primary/workers here). Could you please provide a bit more info on what's the intention here?

File: crates/sui-node/src/lib.rs
Comment (akichidis) [inline crates/sui-node/src/lib.rs:471]: Probably not a comment related that much to this PR, but as an overall observation, we might need to consider the reconfiguration case with committee change. I am not aware what's the plan for networking around this (ex shutting down network? adding/removing peers from the `p2p_network` ? ) , because the connectivity monitor might also need to get updated. 
CC @bmwill

File: narwhal/network/src/connectivity.rs
Comment (akichidis) [inline narwhal/network/src/connectivity.rs:31]: Having an `mpsc sender` here seems to me quite restrictive an tailoring the solution for our case. I would suggest if you want to stick to that implementation, then changing to a [broadcast](https://docs.rs/tokio/latest/tokio/sync/broadcast/) instead might allow more to consume the connectivity status and let the `ConnectionMonitor` play a wider role in the codebase

File: narwhal/network/src/connectivity.rs
Comment (akichidis) [inline narwhal/network/src/connectivity.rs:104]: You probably want to put this outside of the outer `if` statement, otherwise we'll attempt to use the `sender` only if `peer_id` is found in the `peer_id_types` in the `map` , which might not be the case (ex we can totally pass in an empty `peer_id_types` map because we don't have tags to associate with , but we might still want to propagate the connectivity status)

File: narwhal/network/src/connectivity.rs
Comment (akichidis) [inline narwhal/network/src/connectivity.rs:117]: Same comment as above

File: crates/sui-core/src/consensus_adapter.rs
Comment (akichidis) [inline crates/sui-core/src/consensus_adapter.rs:628]: As an extra optimisation - which we didn't have before (so we might do it on follow up PR) - we can probably pass in the `process_waiter` and do a `select` between the `process_waiter` and the `timeout`. If the `process_waiter` has returned successfully, then there is no need for us to keep waiting and trying to submit in consensus. Or alternatively we can even do this in a higher level (when we execute the `submit_after_selected` method)

File: crates/sui-core/src/consensus_adapter.rs
Comment (akichidis) [inline crates/sui-core/src/consensus_adapter.rs:342]: I've left a comment bellow on another part of the code, but this is important to stay here as it measures critical end to end latencies which we won't be to acquire otherwise

File: crates/sui-core/src/consensus_adapter.rs
Comment (akichidis) [inline crates/sui-core/src/consensus_adapter.rs:197]: Similar to my other comment bellow about the epoch change with committee change - unless we create a new `ConnectionMonitorStatus` per epoch, this map will never get cleared up with members that leave. It _might be_ for the beginning assuming that validators won't change a lot so we won't expect this map to grow, but on the long run I don't feel comfortable leaving this unbounded.

File: crates/sui-core/src/consensus_adapter.rs
Comment (akichidis) [inline crates/sui-core/src/consensus_adapter.rs:202]: Could you please add some test coverage for this?

**Review by akichidis — COMMENTED** [review]
Thanks a lot @laura-makdah for this! I would like to do a further pass on it , but please find some comments for now - happy to talk offline if needed to.

I would just like to confirm the logic or re-submission when the previous submitter has failed because we do evaluate the connectivity before we try to initially submit, but never afterwards and we preserve no back off capabilities to ensure we won't deal with spikes (all validators try to re-submit at same time).

File: crates/sui-node/src/lib.rs
Comment (andll) [inline crates/sui-node/src/lib.rs:452]: Unfortunately it is not safe to use `get_sui_system_state_object` in this place as it does not guarantee to hold system state for current epoch.

Instead, to expose information you are lookin for, we could do the following:

(1) There is `EpochStartConfigration` - you can add fields that you are interested in to this object.
(2) You can find a code where it is initialized and pass more data from `SuiSystemObject` into `EpochStartConfigration`. This is the only place in code where access to `SuiSystemObject` is safe.
(3) Here in SuiNode:: construct_validator_components you can use `epoch_store.get_epoch_start_configuration` safely to get the data that you need.

File: crates/sui-types/src/sui_system_state.rs
Comment (andll) [reply crates/sui-types/src/sui_system_state.rs:296]: More reasonable message is probably something like `"Malformed network public key"`.

File: crates/sui-types/src/sui_system_state.rs
Comment (andll) [inline crates/sui-types/src/sui_system_state.rs:286]: Seems like this method should return `HashMap<PeerId, SomeEnum>`, rather then a String with pre-defined values like `"other_primary"`, `"other_worker"`, etc.

File: crates/sui-types/src/sui_system_state.rs
Comment (andll) [inline crates/sui-types/src/sui_system_state.rs:321]: I think you can just use a regular pattern to collect iterator of pairs into a HashMap, instead of creating a temporary `mut result`:
```
return self.validators
            .active_validators
            .iter()
            .map(|validator| {
...
(peer_id, name)
}).collect();
```

File: crates/sui-core/src/consensus_adapter.rs
Comment (andll) [inline crates/sui-core/src/consensus_adapter.rs:129]: In terms of this file organization, can we actually keep the order of things as 

* ConsensusAdapter struct definition
* Other non-test structs definition
* ConsensusAdapter implementation
* Other structs implementation
* Test structs definition
* Test structs implementation

This is roughly how we keep other files in sui - main things on top of the file, additional things down below

File: crates/sui-core/src/consensus_adapter.rs
Comment (andll) [reply crates/sui-core/src/consensus_adapter.rs:226]: +1 to removing async from this method (and from ConsensusAdapter::spawn that we added in this PR)

File: crates/sui-core/src/consensus_adapter.rs
Comment (andll) [inline crates/sui-core/src/consensus_adapter.rs:250]: We need some story to understand when does this task finish (to make sure it actually completes on epoch boundary).
Is it guaranteed that `self.receiver` is dropped when epoch ends? Why? 
(Can you reply to this question as a comment to the `run` method as this is worth mentioning in the code itself?)

File: crates/sui-core/src/consensus_adapter.rs
Comment (andll) [inline crates/sui-core/src/consensus_adapter.rs:544]: Can we try to retain as much of existing logic as possible with the change?

* We can keep the order and `wait_submit` logic
* Limit to only removing the disconnected peers from the list before calculating our own position

**Review by andll — CHANGES_REQUESTED** [review]
Added some comments

Most critical parts are the following
(1) We need to change how we access SuiSystemObject at the sui node start - I added some details on the safe way to do it
(2) I think we can limit size of this change by only filtering out the disconnected peers from the list. I think other things can stay intact.

**Comment by andll** [comment]
Another thing to note is I think you can also simplify this PR by simply updating your `dashmap` in the `ConnectionManager`, instead of spawning the task and and having a communication channel. This way we don't need to reason about when the monitoring task will be finished, and it will further reduce the PR size (will remove few unnecessary structs)

File: crates/sui-node/src/lib.rs
Comment (bmwill) [reply crates/sui-node/src/lib.rs:471]: Yeah we'll need to, at least for validators, update the known peers in order to ensure validators in the new epoch have whitelisted each other

**Comment by andll** [comment]
For accessing SuiSystemState from EpochStartConfiguration, https://github.com/MystenLabs/sui/pull/8366 will add SuiSystemState there for you

File: crates/sui-types/src/sui_system_state.rs
Comment (laura-makdah) [reply crates/sui-types/src/sui_system_state.rs:323]: I wasn't able to make this work but tried

File: crates/sui-core/src/consensus_adapter.rs
Comment (laura-makdah) [reply crates/sui-core/src/consensus_adapter.rs]: I would like to note that after a second look, the original code has a max sleep time of 10 seconds, and since we add 7s seconds as a multiplier to each position, if we had (and with updates to this PR) currently have a latency of 10 seconds or more, we will submit from every authority. I've changed it to remain how it is, but its worth keeping in mind as we evaluate going forward.

File: crates/sui-core/src/consensus_adapter.rs
Comment (laura-makdah) [reply crates/sui-core/src/consensus_adapter.rs]: I've now removed this

**Comment by laura-makdah** [comment]
@andll @akichidis Thanks for the thorough reviews, evaluating them as a whole I have made the following major updates
• removed the ConsensusMonitorListener component entirely, which was there to offload an extra map lookup on the critical path of each tx. It was probably not worth the complication. 
• Instead of creating a new ConnectionMonitor or needing to update it, it runs all the time, with the DashMap that gets injected into the ConsensusAdapter (which will only exist some epochs as a node gets upgraded or downgraded) getting saved on SuiNode struct. This gets updated with new epoch-specific mapping every epoch, regardless of validator status. 
• I have retained the framework of the original submission logic that calculates the sleep time for each node as a function of position, and with it reverted back to the original metrics initialization 

For the remaining comments, most of them became inapplicable, for the remaining comments that still apply they are in the final commit.

File: crates/sui-core/src/consensus_adapter.rs
Comment (akichidis) [reply crates/sui-core/src/consensus_adapter.rs]: @laura-makdah could you please point the code where we do have the max sleep time of 10 seconds - just want to make sure we are talking about the same thing. When looking to the original `submit_delay_certificate` method, I can see that the maximum delay multiplier is 10, so the maximum backoff time will be 10*7 = 70 seconds , meaning that any validator after the position 10 will have waited for maximum 70 seconds, and then yes, all of them would go and submit the transaction.

Comment (laura-makdah) [reply crates/sui-core/src/consensus_adapter.rs]: Ah, you're right

File: narwhal/network/src/connectivity.rs
Comment (akichidis) [reply narwhal/network/src/connectivity.rs:22]: If that is not really an option/don't like it, then what about injecting the `ConnectionMonitorStatus` directly? I think that would make the intention clearer and not having do deal with components that act as bridges (in this case the `connection_statuses` map)

**Review by akichidis — COMMENTED** [review]
Great work @laura-makdah  - I like the simplification compared to the previous approach! I've left a few comments, let me know your thoughts. Also, could you please add some testing around the edge cases (ex validators are disconnected)

Also, I would highly recommend deploy this in private-testnet and test it before we rollout in main as it is sensitive change with some testing plan (ex bringing down validators and monitor specific metrics to confirm the behaviour / benefit from it)

File: crates/sui-node/src/lib.rs
Comment (andll) [inline crates/sui-node/src/lib.rs:676]: Why is this `async`?

File: crates/sui-core/src/consensus_adapter.rs
Comment (andll) [inline crates/sui-core/src/consensus_adapter.rs:433]: Can we remove this for now to make this PR cleaner? When we introduce narwhal scores we can always add this parameter

File: crates/sui-core/src/consensus_adapter.rs
Comment (andll) [reply crates/sui-core/src/consensus_adapter.rs:364]: Yeah it would be good to understand why this won't fail (e.g. what is the guarantee that check_connection always returns true for ourselves).
Ideally we can just modify the code so that we don't rely on this property for `check_connection` - perhaps always return `Connected` for our peer or do not call `check_connection` for ourselves above

File: crates/sui-core/src/consensus_adapter.rs
Comment (laura-makdah) [reply crates/sui-core/src/consensus_adapter.rs:357]: I considered this actually and that is slightly different. What that would acheive is for all the low scoring nodes to remove all the other low scoring nodes but themselves, which would result in more of the low scoring nodes to submit the same transaction sometimes.

Comment (akichidis) [reply crates/sui-core/src/consensus_adapter.rs:357]: > What that would acheive is for all the low scoring nodes to remove all the other low scoring nodes but themselves

Ok but if we have the code like this:
```
!low_scoring_authorities.contains(&authority)
                    || low_scoring_authorities.contains(&ourselves)
```

then if _we_ are a low scoring authority, we'll always include everyone else in the final set - irrespective of whether we consider them low scoring authority or not - so practically it would be like disabling the scoring mechanism, right? Is this intended?

File: crates/sui-core/src/consensus_adapter.rs
Comment (laura-makdah) [reply crates/sui-core/src/consensus_adapter.rs:336]: It still is the same algorithm, and the result will be the same, (skipping over and "removing" are essentially the same behavior) but with the addition of the low performing co-submits described at 334-336 section

File: .cargo/config
Comment (laura-makdah) [reply .cargo/config:1]: re-added

File: crates/sui-node/src/lib.rs
Comment (laura-makdah) [reply crates/sui-node/src/lib.rs:676]: removed

File: crates/sui-core/src/consensus_adapter.rs
Comment (laura-makdah) [reply crates/sui-core/src/consensus_adapter.rs:433]: I have already made the commitment to including this functionality during the planning meeting

Comment (andll) [reply crates/sui-core/src/consensus_adapter.rs:437]: This is not a question about commitment - let's keep PRs limited to one change at a time. There is no problem with adding low-score related things in the future PRs. Mixing in random pars of future PRs makes things more complicated to review and raise questions that are hard to answer without seeing the future PR(like for example, why is low_scoring_authorities defined here? Why not abstract all of this out of consensus adapter into some kind of `ExcludeAuthority` trait.

File: crates/sui-core/src/consensus_adapter.rs
Comment (laura-makdah) [reply crates/sui-core/src/consensus_adapter.rs:361]: Yes, that is intentional, in the case a tx gets mapped like this:
[low_scoring_a1, low_scoring_a2, a3]
It would result in low_scoring_a1 and a3 to submit the tx since low_scoring_a2 did not filter out low_scoring_a1. I was aiming to avoid an amplification where all 3 submit the same tx.

**Comment by laura-makdah** [comment]
@andll @akichidis this is ready for re-review

**Comment by laura-makdah** [comment]
part of test plan: https://github.com/MystenLabs/sui-operations/pull/1217

**Comment by andll** [comment]
I still think we should remove code for low_scoring_authorities to focus on one thing in this PR(see my comment here: https://github.com/MystenLabs/sui/pull/7972#discussion_r1119125478)

If you want to keep it for whatever reason, then we probably need to look how to encapsulate this outside of ConsensusAdapter

**Comment by laura-makdah** [comment]
@andll Ok, I've removed it from this PR

File: crates/sui-core/src/consensus_adapter.rs
Comment (andll) [inline crates/sui-core/src/consensus_adapter.rs:259]: Because you now added `self` to this function you can remove `ourselves` argument (it can be read directly from `self`)

File: crates/sui-core/src/consensus_adapter.rs
Comment (andll) [inline crates/sui-core/src/consensus_adapter.rs:289]: Same, `ourselves` can be removed

File: crates/sui-core/src/consensus_adapter.rs
Comment (andll) [inline crates/sui-core/src/consensus_adapter.rs:592]: Is this function used anywhere outside of tests? If no, we should remove it (and update tests to use some other function that we made perhaps?)

File: crates/sui-core/src/consensus_adapter.rs
Comment (andll) [inline crates/sui-core/src/consensus_adapter.rs:320]: Same, `ourselves` can be read from `self` directly

File: crates/sui-core/src/consensus_adapter.rs
Comment (andll) [inline crates/sui-core/src/consensus_adapter.rs:324]: I think it would be more obvious if we just do check for `ourselves` in this code block, instead of passing it to the `check_connection`.
E.g. `ourselves` can be removed from `check_connection` and this line can be rewritten as following so that it is perfectly clear the `expect` below does not panic:

```let filtered_positions = positions.into_iter().filter(|authority| {
            ourselves == authority || // Always include ourselves
            self.connection_monitor_status
                .check_connection(authority) == Some(ConnectionStatus::Connected)
});
```

File: crates/sui-core/src/consensus_adapter.rs
Comment (andll) [inline crates/sui-core/src/consensus_adapter.rs:284]: Why did you change this comment to `1 or 2`? I think logic in this PR still selects one validator only?

**Review by andll — COMMENTED** [review]
Looks pretty good now, but left few more comments

File: crates/sui-core/src/consensus_adapter.rs
Comment (laura-makdah) [reply crates/sui-core/src/consensus_adapter.rs:284]: In the case of a network partition, it will select 2

File: crates/sui-core/src/consensus_adapter.rs
Comment (laura-makdah) [reply crates/sui-core/src/consensus_adapter.rs:324]: ok, I am adding this but also keeping the other check for safety

File: crates/sui-core/src/consensus_adapter.rs
Comment (laura-makdah) [reply crates/sui-core/src/consensus_adapter.rs]: This was and still is testing the order_validators_for_sumbmission function, I've moved its location to become clearly a test helper function

File: crates/sui-core/src/consensus_adapter.rs
Comment (andll) [reply crates/sui-core/src/consensus_adapter.rs:284]: Can you explain this? I see the code and only one validator receives position 0, where does second transactio comes from in normal case?
Mingwei made a change in testnet branch that would result in multiple validators submitting transaction, but this is not part of main branch or this PR

File: crates/sui-core/src/consensus_adapter.rs
Comment (andll) [reply crates/sui-core/src/consensus_adapter.rs]: But this is exactly my concern - we used to use `position_submit_certificate` in the code, and we wrote unit tests for it.

Now you switched to using `submission_position` function in the prod tests, but unit test continue to use `position_submit_certificate`. Don't you think unit tests also need to be updated to test same function(`submission_position`) that production code uses??

Comment (laura-makdah) [reply crates/sui-core/src/consensus_adapter.rs]: Ok, so you are asking for a unit test for the function `submission_possiotion`, or more specifically `check_submission_wrt_connectivity`? I still see the random ordering of validators, or rather the functionality in `order_validators_for_submission` to be the part that needs testing, since without that part working as expected, we may have no-one submit a tx.  The functionality of removing authorities from a list doesn't warrant a unit test, in my opinion, so I didn't add one. Also I didn't add a unit test for checking connectivity, because we rely on the connectivity monitor in our dashboards, I did assume whoever created those has verified their correctness.

Comment (andll) [reply crates/sui-core/src/consensus_adapter.rs]: Is it hard to switch unit test to use `submission_position` and pass your `ConnectivityStatusForTest` to consensus adapter?
The reason we prefer to have unit tests for the functions that are actually used in real code is that when something change in those functions, we want to get signal from unit tests.

**Review by andll — APPROVED** [review]
We should try to test this in private testnet and see that TPS does not change, but otherwise seems good

I would prefer to migrate unit test to use `submission_position` (since this is what prod code is using now), but not blocking on it

File: crates/sui-core/src/consensus_adapter.rs
Comment (akichidis) [inline crates/sui-core/src/consensus_adapter.rs:515]: Could you please add a test for this?

File: crates/sui-core/src/consensus_adapter.rs
Comment (akichidis) [inline crates/sui-core/src/consensus_adapter.rs:312]: Could you please add some testing for this method?

**Review by akichidis — APPROVED** [review]
Added a couple of more comments but this is great work @laura-makdah !
I would highly suggest adding a few tests though for the unhappy path (when nodes are disconnected) so we can confirm the expected behaviour. I am approving to unblock since there is a testing plan for private-testnet, but I if we could increase the test coverage here (even on follow up PR) I would feel safer.

File: crates/sui-benchmark/tests/simtest.rs
Comment (mystenmark) [inline crates/sui-benchmark/tests/simtest.rs:149]: please just add `#[ignore]` instead of commenting the test out.

**Pull request merged**
